internal/trace/listener.go

```diff
@@ -11,19 +11,144 @@ package trace
 import (
 	"context"
 	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/DataDog/datadog-lambda-go/internal/logger"
+	"github.com/aws/aws-lambda-go/lambdacontext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
 )
 
 type (
-	// Listener implements HandlerListener, injecting datadog tracing info into the context
-	Listener struct{}
+	// Listener creates a function execution span and injects it into the context
+	Listener struct {
+		DDTraceEnabled  bool
+		MergeXrayTraces bool
+	}
+
+	// Config gives options for how the Listener should work
+	Config struct {
+		DDTraceEnabled  bool
+		MergeXrayTraces bool
+	}
 )
 
-// HandlerStarted adds trace metadata to the context
+// The function execution span is the top-level span representing the current Lambda function execution
+var functionExecutionSpan ddtrace.Span
+
+// MakeListener initializes a new trace lambda Listener
+func MakeListener(config Config) Listener {
+
+	return Listener{
+		DDTraceEnabled:  config.DDTraceEnabled,
+		MergeXrayTraces: config.MergeXrayTraces,
+	}
+}
+
+// HandlerStarted creates the function execution span representing the Lambda function execution
+// and adds that span to the context so that the user can create child spans (if Datadog tracing is enabled)
 func (l *Listener) HandlerStarted(ctx context.Context, msg json.RawMessage) context.Context {
-	ctx, _ = ExtractTraceContext(ctx, msg)
+	ctx, _ = ContextWithTraceContext(ctx, msg)
+
+	if l.DDTraceEnabled {
+		tracer.Start(
```

Comment:
What happens when tracer.Start gets called multiple times? Does it trash any existing state, or try to reinstrument, or is it a no-op?

Reply:
Per the `dd-trace-go` docs:

> It will stop and replace any running tracer, meaning that calling it several times will result in a restart of the tracer by replacing the current instance with a new one.

However, I am stopping the trace after every invocation anyway.